File: src/clojure/convert_tests.clj
#!/usr/bin/env clojure
;;; Script to convert .techne CommonLisp test files to Clojure

(require '[clojure.string :as str])
(require '[clojure.java.io :as io])

(defn convert-line [line]
  "Convert a single line from CommonLisp to Clojure syntax"
  (cond
    ;; Comment lines stay the same
    (str/starts-with? line ";;") line

    ;; Skip load statements
    (str/includes? line "(load ") ""

    ;; Convert defvar to def
    (str/includes? line "(defvar *rekb*")
    "(def rekb (create-rekb \"Random rekb\"))"

    ;; Convert setf with node declarations - just mark for later processing
    (str/starts-with? (str/trim line) "(setf") :start-setf

    ;; End of setf block
    (and (str/starts-with? (str/trim line) ")")
         (not (str/includes? line "assrt-formula")))
    :end-setf

    ;; Node declarations within setf
    (re-find #"node-\d+.*declare-atomic" line)
    (let [match (re-find #"(node-\d+) \(declare-atomic nil \"(.+?)\" :(\w+) \*rekb\*\)" line)]
      (when match
        (str "(def " (nth match 1) " (declare-atomic nil \"" (nth match 2) "\" :" (nth match 3) " rekb))")))

    ;; Convert assrt-formula calls
    (str/includes? line "assrt-formula")
    (-> line
        (str/replace "(assrt-formula " "(assrt-formula ")
        (str/replace "(list " "[")
        (str/replace "*rekb*)" "rekb)")
        (str/replace " *rekb*)" " rekb)")
        (str/replace #"\)(?=\s*;)" "]")
        (str/replace #"\)\s*$" "")
        (str/replace #"\s+$" ""))

    ;; Empty lines
    (str/blank? line) ""

    ;; Default: return as-is
    :else line))

(defn process-setf-block [lines]
  "Process the setf block to extract node definitions"
  (let [node-lines (filter #(re-find #"node-\d+" %) lines)]
    (map (fn [line]
           (let [match (re-find #"(node-\d+) \(declare-atomic nil \"(.+?)\" :(\w+) \*rekb\*\)" line)]
             (when match
               (str "(def " (nth match 1) " (declare-atomic nil \"" (nth match 2) "\" :" (nth match 3) " rekb))"))))
         node-lines)))

(defn convert-file [input-path output-path]
  "Convert a .techne file to Clojure"
  (println (str "Converting " input-path " to " output-path))
  (let [lines (str/split-lines (slurp input-path))
        in-setf (atom false)
        setf-lines (atom [])]
    (with-open [writer (io/writer output-path)]
      ;; Write header
      (.write writer ";; ")
      (.write writer (first lines))
      (.write writer "\n")
      (.write writer "(require '[techne-psm :refer :all])\n\n")
      (.write writer "(def rekb (create-rekb \"Random rekb\"))\n\n")

      ;; Process remaining lines
      (doseq [line (rest lines)]
        (let [result (convert-line line)]
          (cond
            (= result :start-setf)
            (reset! in-setf true)

            (= result :end-setf)
            (do
              ;; Process accumulated setf lines
              (doseq [def-line (process-setf-block @setf-lines)]
                (when def-line
                  (.write writer def-line)
                  (.write writer "\n")))
              (.write writer "\n")
              (reset! in-setf false)
              (reset! setf-lines []))

            @in-setf
            (swap! setf-lines conj line)

            (and (string? result) (not (str/blank? result)))
            (do
              (.write writer result)
              (.write writer "\n"))

            :else nil))))))

;; Main execution
(defn -main []
  (let [examples-dir "examples"
        output-dir "examples/clojure"
        techne-files (filter #(str/ends-with? (.getName %) ".techne")
                            (file-seq (io/file examples-dir)))]

    ;; Skip rand-50.techne since we already converted it manually
    (doseq [file techne-files]
      (let [filename (.getName file)
            basename (str/replace filename #"\.techne$" "")]
        (when (not= basename "rand-50")
          (convert-file (.getPath file)
                       (str output-dir "/" basename ".clj")))))

    (println "\nConversion complete!")))

;; Run if executed as script
(when (= *file* (System/getProperty "babashka.file"))
  (-main))

File: examples/convert_tests.py
#!/usr/bin/env python3
"""Script to convert .techne CommonLisp test files to Clojure"""

import re
import os
from pathlib import Path

def convert_techne_to_clojure(input_file, output_file):
    """Convert a single .techne file to Clojure"""

    with open(input_file, 'r') as f:
        lines = f.readlines()

    output_lines = []
    in_setf = False
    node_defs = []

    for i, line in enumerate(lines):
        # Keep comment lines
        if line.strip().startswith(';;'):
            if i == 0:
                output_lines.append(line)
            continue

        # Skip load statements
        if '(load ' in line:
            continue

        # Skip defvar line
        if 'defvar *rekb*' in line:
            continue

        # Start of setf block
        if line.strip().startswith('(setf') or '(setf' in line:
            in_setf = True
            # Check if first node is on same line as setf
            match = re.search(r'(node-\d+)\s+\(declare-atomic nil "(.+?)" :(\w+) \*rekb\*\)', line)
            if match:
                node_name, label, sort = match.groups()
                node_defs.append(f'(def {node_name} (declare-atomic nil "{label}" :{sort} rekb))\n')
            continue

        # End of setf block (lone closing paren)
        if in_setf and line.strip() == ')':
            in_setf = False
            continue

        # Collect node definitions within setf
        if in_setf:
            match = re.search(r'(node-\d+)\s+\(declare-atomic nil "(.+?)" :(\w+) \*rekb\*\)', line)
            if match:
                node_name, label, sort = match.groups()
                node_defs.append(f'(def {node_name} (declare-atomic nil "{label}" :{sort} rekb))\n')
            continue

        # Convert assrt-formula lines
        if 'assrt-formula' in line:
            # Replace list with vector syntax
            converted = line.replace('(list ', '[')
            # Replace *rekb* with rekb
            converted = converted.replace('*rekb*', 'rekb')
            # Fix closing parens for lists
            converted = re.sub(r'\)(\s+:DA)', r']\1', converted)
            output_lines.append(converted)

    # Write output file
    with open(output_file, 'w') as f:
        # Write header
        f.write(output_lines[0])  # Comment line
        f.write("(require '[techne-psm :refer :all])\n\n")
        f.write('(def rekb (create-rekb "Random rekb"))\n\n')

        # Write node definitions
        for node_def in node_defs:
            f.write(node_def)
        f.write('\n')

        # Write assrt-formula calls
        for line in output_lines[1:]:
            f.write(line)

def main():
    examples_dir = Path('examples')
    output_dir = Path('examples/clojure')

    # Find all .techne files
    techne_files = list(examples_dir.glob('*.techne'))

    print(f"Found {len(techne_files)} .techne files to convert")

    for techne_file in techne_files:
        # Skip rand-50 since we already converted it manually
        if techne_file.stem == 'rand-50':
            print(f"Skipping {techne_file.name} (already converted)")
            continue

        output_file = output_dir / f"{techne_file.stem}.clj"
        print(f"Converting {techne_file.name} -> {output_file.name}")

        try:
            convert_techne_to_clojure(techne_file, output_file)
        except Exception as e:
            print(f"  Error: {e}")

    print("\nConversion complete!")

if __name__ == '__main__':
    main()
